feat: CA bundle mount options for storage initializer (#3250)
kserve/kserve: docs/samples/cabundle/README.md

> # KServe with Self Signed Certificate Model Registry

Reviewer: It is probably better to move these docs to the website for more visibility: https://github.com/kserve/website

Author: yes good idea

This document explains three methods that can be used in KServe.

- Configure CaBundle for storage-initializer
  - Global configuration
  - Using `storage-config` Secret

- Skip SSL Verification
  
## Configure CaBundle for storage-initializer  
### Global Configuration

KServe use `inferenceservice-config` ConfigMap for default configuration. If you want to add `cabundle` cert for every inferenceservice, you can set `caBundleSecretName` in the ConfigMap. Before updating the ConfigMap, you have to create a secret for cabundle certificate in the namespace that KServe controller is running and the data key in the Secret must be cabundle.crt. 

- Create a Secret with the cabundle cert
  ~~~
  kubectl create secret generic cabundle --from-file=/path/to/cabundle.crt

  kubectl get secret cabundle -o yaml
  apiVersion: v1
  data:
    cabundle.crt: LSXXXK
  kind: Secret
  metadata:
    name: cabundle
    namespace: kserve
  type: Opaque
  ~~~
- Update `inferenceservice-config` ConfigMap 
  ~~~
    storageInitializer: |-
    {
        ...
        "caBundleSecretName": "cabundle",
        ...
    }
  ~~~
  
If you update this configuration after, please restart KServe controller pod.  

> ### Using storage-config Secret

Reviewer: Would this support the [common storage secret](https://github.com/kserve/kserve/tree/master/docs/samples/storage/storageSpec#create-common-storage-secret) format?

If so, for the common storage we need either (1) a way to attach multiple bundles to the single secret, or (2) be able to specify a bundle for each entry in the single secret. But that is not mentioned here.

Author: No, it is a special option so annotation would be enough.

If you want to apply the cabundle only to a specific inferenceservice, you can use a specific annotation on the `storage-config` Secret used by the inferenceservice.
In this case, you have to create the cabundle Secret in the user namespace before you create the inferenceservice.

- Create a Secret with the cabundle cert
  ~~~
  kubectl create secret generic local-cabundle --from-file=/path/to/cabundle.crt

  kubectl get secret cabundle -o yaml
  apiVersion: v1
  data:
    cabundle.crt: LSXXXK
  kind: Secret
  metadata:
    name: local-cabundle
    namespace: kserve
  type: Opaque
  ~~~

- Add an annotation to `storage-config` Secret
  ~~~
  apiVersion: v1
  data:
    AWS_ACCESS_KEY_ID: VEhFQUNDRVNTS0VZ
    AWS_SECRET_ACCESS_KEY: VEhFUEFTU1dPUkQ=
  kind: Secret
  metadata:
    annotations:
      serving.kserve.io/s3-cabundle-secret: local-cabundle
      ...
    name: storage-config
    namespace: kserve-demo
  type: Opaque
  ~~~

> ## Skip SSL Verification

Reviewer: Also here, would this support the [common storage secret](https://github.com/kserve/kserve/tree/master/docs/samples/storage/storageSpec#create-common-storage-secret) format?

The ability to skip per entry, or globally in the single secret would also be needed.

Author: I don't know if it is possible globally. @yuzisun do you have any thoughts on this?

Author: this will be solved by this pr(https://github.com/kserve/kserve/pull/3267)

# KServe with Self Signed Certificate Model Registry

> If you are using a model registry with a self-signed certificate, you must either skip ssl verify or apply the appropriate cabundle to the storage-initializer to create a connection with the registry.

Reviewer 2: ```suggestion
If you are using a model registry with a self-signed certificate, you must either skip ssl verify or apply the appropriate CA bundle to the storage-initializer to create a connection with the registry.
```

Author: updated

This document explains three methods that can be used in KServe, described below:

- Configure CA bundle for storage-initializer
  - Global configuration
  - Using `storage-config` Secret

- Skip SSL Verification
  
## Configure CaBundle for storage-initializer  
### Global Configuration

KServe use `inferenceservice-config` ConfigMap for default configuration. If you want to add `cabundle` cert for every inference service, you can set `caBundleConfigMapName` in the ConfigMap. Before updating the ConfigMap, you have to create a ConfigMap for CA bundle certificate in the namespace that KServe controller is running and the data key in the ConfigMap must be `cabundle.crt`. 

- Create a ConfigMap with the CA bundle cert
  ~~~
  kubectl create configmap cabundle --from-file=/path/to/cabundle.crt

  kubectl get configmap cabundle -o yaml
  apiVersion: v1
  data:

>     cabundle.crt: XXXXX

Reviewer 2: nit: Kubernetes uses key ca.crt

Author: that's right bit I tried to use a slightly different name so as not to overlap with anything else.

  kind: ConfigMap
  metadata:
    name: cabundle
    namespace: kserve
  ~~~
- Update `inferenceservice-config` ConfigMap 
  ~~~
    storageInitializer: |-
    {
        ...
        "caBundleConfigMapName": "cabundle",
        ...
    }
  ~~~
  
If you update this configuration after, please restart KServe controller pod.  

### Using storage-config Secret

If you want to apply the cabundle only to a specific inferenceservice, you can use a specific annotation or variable(`cabundle_configmap`) on the `storage-config` Secret used by the inferenceservice.
In this case, you have to create the cabundle ConfigMap in the user namespace before you create the inferenceservice.

- Create a ConfigMap with the cabundle cert
  ~~~
  kubectl create configmap local-cabundle --from-file=/path/to/cabundle.crt

  kubectl get configmap cabundle -o yaml
  apiVersion: v1
  data:
    cabundle.crt: XXXXX
  kind: ConfigMap
  metadata:
    name: local-cabundle

>     namespace: kserve

Reviewer 2: this should be kserve-demo namespace?

Author: updated.

# KServe with Self Signed Certificate Model Registry

If you are using a model registry with a self-signed certificate, you must either skip ssl verify or apply the appropriate cabundle to the storage-initializer to create a connection with the registry.

> This document explains three methods that can be used in KServe.

Reviewer 3: ```suggestion
This document explains three methods that can be used in KServe, described below:
```

Author: updated

> - Configure CaBundle for storage-initializer

Reviewer 3: ```suggestion
- Configure CA bundle for storage-initializer
```

Author: updated

  - Global configuration
  - Using `storage-config` Secret

- Skip SSL Verification
  
## Configure CaBundle for storage-initializer  
### Global Configuration

> KServe use `inferenceservice-config` ConfigMap for default configuration. If you want to add `cabundle` cert for every inferenceservice, you can set `caBundleConfigMapName` in the ConfigMap. Before updating the ConfigMap, you have to create a configmap for cabundle certificate in the namespace that KServe controller is running and the data key in the ConfigMap must be cabundle.crt. 

Reviewer 3: ```suggestion
KServe use `inferenceservice-config` ConfigMap for default configuration. If you want to add `cabundle` cert for every inference service, you can set `caBundleConfigMapName` in the ConfigMap. Before updating the ConfigMap, you have to create a ConfigMap for CA bundle certificate in the namespace that KServe controller is running and the data key in the ConfigMap must be cabundle.crt. 
```

Author: updated

> - Create a ConfigMap with the cabundle cert

Reviewer 3: ```suggestion
- Create a ConfigMap with the CA bundle cert
```

Author: updated

  ~~~
  kubectl create configmap cabundle --from-file=/path/to/cabundle.crt

  kubectl get configmap cabundle -o yaml
  apiVersion: v1
  data:
    cabundle.crt: XXXXX
  kind: ConfigMap
  metadata:
    name: cabundle
    namespace: kserve
  ~~~
- Update `inferenceservice-config` ConfigMap 
  ~~~
    storageInitializer: |-
    {
        ...
        "caBundleConfigMapName": "cabundle",
        ...
    }
  ~~~
  
If you update this configuration after, please restart KServe controller pod.  

### Using storage-config Secret

If you want to apply the cabundle only to a specific inferenceservice, you can use a specific annotation on the `storage-config` Secret used by the inferenceservice.
In this case, you have to create the cabundle ConfigMap in the user namespace before you create the inferenceservice.

> - Create a Secret with the cabundle cert

Reviewer: ```suggestion
- Create a ConfigMap with the cabundle cert
```

  ~~~
  kubectl create configmap local-cabundle --from-file=/path/to/cabundle.crt

  kubectl get configmap cabundle -o yaml
  apiVersion: v1
  data:
    cabundle.crt: XXXXX
  kind: ConfigMap
  metadata:
    name: local-cabundle
    namespace: kserve
  ~~~

- Add an annotation to `storage-config` Secret
  ~~~

>   apiVersion: v1

Reviewer 3: if this secret is already created, what about to add a oc/kubectl patch apply command?

Author: no it will be created by users.
